File: Python/ByTier/B/2046_B_Move_Back_at_a_Cost.py
# Problem: CF 2046 B - Move Back at a Cost
# https://codeforces.com/contest/2046/problem/B

"""
B. Move Back at a Cost
Algorithm:Greedy with preprocessing
Time Complexity: O(n log n) due to sorting the P array
Space Complexity: O(n) for storing SUFMIN, P, and N arrays

The problem asks to find the lexicographically smallest array by performing operations:
- Choose an element, increment it by 1, and move it to the back.
This simulates a greedy process where we want to push larger elements towards the end 
to minimize the prefix of the final result.

Approach:
1. Preprocess suffix minimums to know the smallest value from each index to the end.
2. Greedily decide whether to increment and move an element or keep it.
3. Use two lists P (elements to be incremented and moved) and N (elements to be kept).
4. Sort P in ascending order, then append it to N to form the final array.
"""

import sys

input = sys.stdin.readline
for _ in range(int(input())):
    n = int(input())
    A = list(map(int, input().split()))
    A.reverse()  # Reverse to process from right to left
    SUFMIN = [int(1e10)]  # Suffix minimum array initialized with large value
    for i in range(n):
        SUFMIN.append(min(SUFMIN[-1], A[i]))  # Build suffix minimums
    P = []  # Elements to be incremented and moved to back
    N = []  # Elements to be kept in place
    omi = int(1e10)  # Holds the minimum of incremented values so far that are pushed back
    for i in range(n - 1, -1, -1):  # Traverse from right to left
        if A[i] > SUFMIN[i] or A[i] > omi:  # If current element is greater than suffix min or pushed-back value
            omi = min(omi, A[i] + 1)  # Update the minimum of future elements that will be pushed_back
            P.append(A[i] + 1)  # Increment and push to back (to P)
        else:
            N.append(A[i])  # Keep in place (to N)
    P.sort()  # Sort increment operations to get lexicographically minimal result
    N += P  # Append sorted P to N
    print(*N)


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/B/2046_B_Move_Back_at_a_Cost.cpp
// Problem: CF 2046 B - Move Back at a Cost
// https://codeforces.com/contest/2046/problem/B

/*
B. Move Back at a Cost
Algorithms/Techniques: Greedy, Simulation, Sorting
Time Complexity: O(n log n) per test case due to sorting
Space Complexity: O(n) for the array and auxiliary arrays

The problem asks to find the lexicographically smallest array by performing operations where we can increase an element and move it to the back.
This greedy approach processes from right to left, increasing elements when necessary to ensure non-decreasing order,
then adjusts remaining elements to minimize the final result.
*/

#include <algorithm>
#include <iostream>

using namespace std;
int t, n, a[100005], mn, mnn, f[100005];
int main() {
  ios::sync_with_stdio(false);
  cin.tie(0), cout.tie(0);
  cin >> t;
  while (t--) {
    cin >> n;
    mn = mnn = 2e9; // Initialize min values to large number
    for (int i = 0; i < n; i++) {
      cin >> a[i];
      f[i] = 0; // Mark array to track processed elements
    }
    for (int i = n - 1; i >= 0; i--) {
      if (mn > a[i])
        mn = a[i]; // Track minimum from right
      if (a[i] > mn) { // If current element is greater than min seen so far
        a[i]++;       // Increment it
        f[i] = 1;     // Mark that this element was incremented
        mnn = min(mnn, a[i]); // Update global minimum after increment
      }
    }
    for (int i = n - 1; i >= 0; i--)
      if (f[i] == 0 && a[i] > mnn) // For unmarked elements that are still greater than min
        a[i]++; // Increment them to reduce lexicographic value
    sort(a, a + n); // Sort result for final output
    for (int i = 0; i < n; i++)
      cout << a[i] << " ";
    cout << "\n";
  }
  return 0;
}


// https://github.com/VaHiX/codeForces/